src/Common/src/System/Net/SocketAddressPal.Unix.cs

```diff
@@ -10,156 +10,129 @@ namespace System.Net
+
+                default:
+                    Debug.Fail("Unexpected failure in GetAddressFamily");
+                    throw new PlatformNotSupportedException();
             }
         }
 
+        public static unsafe AddressFamily GetAddressFamily(byte[] buffer)
+        {
+            AddressFamily family;
+            Interop.Error err;
+            fixed (byte* rawAddress = buffer)
+            {
+                err = Interop.Sys.GetAddressFamily(rawAddress, buffer.Length, (int*)&family);
+            }
+
+            ThrowOnFailure(err);
+            return family;
+        }
+
         public static unsafe void SetAddressFamily(byte[] buffer, AddressFamily family)
         {
+            Interop.Error err;
             fixed (byte* rawAddress = buffer)
             {
-                var sockaddr = (Interop.libc.sockaddr*)rawAddress;
-                Interop.CheckBounds((byte*)sockaddr, buffer.Length, &sockaddr->sa_family);
-
-                switch (family)
-                {
-                    case AddressFamily.Unspecified:
-                        sockaddr->sa_family = Interop.libc.AF_UNSPEC;
-                        break;
-
-                    case AddressFamily.Unix:
-                        sockaddr->sa_family = Interop.libc.AF_UNIX;
-                        break;
-
-                    case AddressFamily.InterNetwork:
-                        sockaddr->sa_family = Interop.libc.AF_INET;
-                        break;
-
-                    case AddressFamily.InterNetworkV6:
-                        sockaddr->sa_family = Interop.libc.AF_INET6;
-                        break;
-
-                    default:
-                        Debug.Fail("Unsupported addres family");
-                        throw new PlatformNotSupportedException();
-                }
+                err = Interop.Sys.SetAddressFamily(rawAddress, buffer.Length, (int)family);
             }
+
+            ThrowOnFailure(err);
         }
 
         public static unsafe ushort GetPort(byte[] buffer)
         {
             ushort port;
-
+            Interop.Error err;
             fixed (byte* rawAddress = buffer)
             {
-                var sockaddr = (Interop.libc.sockaddr*)rawAddress;
-                switch (sockaddr->sa_family)
-                {
-                    case Interop.libc.AF_INET:
-                        port = Interop.CheckedRead((byte*)sockaddr, buffer.Length, &((Interop.libc.sockaddr_in*)sockaddr)->sin_port);
-                        break;
-
-                    case Interop.libc.AF_INET6:
-                        port = Interop.CheckedRead((byte*)sockaddr, buffer.Length, &((Interop.libc.sockaddr_in6*)sockaddr)->sin6_port);
-                        break;
-
-                    default:
-                        Debug.Fail("Unsupported address family");
-                        throw new PlatformNotSupportedException();
-                }
+                err = Interop.Sys.GetPort(rawAddress, buffer.Length, &port);
```

Comment:
In most cases, byte[] would be a reasonable choice for the majority of people because it is easier concept for most developers, where perf isn't super critical. But it does have its down sides - it is less flexible, pinning makes memory more fragmented and makes GC less efficient, and might lead to extra memory allocation/free in out/ref cases, etc. In potentially perf critical scenarios in our network libraries, I would agree using byte* is probably the better choice here. And there is no perf loss - we are just passing a pointer. Well, in the case of this change, there is no practical difference since you are doing your own pinning.